apps/extension/src/storage/base.ts

```diff
@@ -20,30 +20,53 @@ export interface StorageItem<T> {
 }
 
 type Version = string;
+
+export type MigrationMap<OldState, NewState> = {
+  [K in keyof OldState & keyof NewState]?: (
+    prev: OldState[K],
+    get: <K extends keyof NewState>(key: K) => Promise<NewState[K]>,
+  ) => NewState[K] | Promise<NewState[K]>;
+};
+
 // It is quite difficult writing a generic that covers all migration function kinds.
 // Therefore, the writer of the migration should ensure it is typesafe when they define it.
-// See `migration.test.ts` for an example.
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-type Migration = Record<Version, (a: any) => any>;
-type Migrations<K extends string | number | symbol> = Partial<Record<K, Migration>>;
+export type Migrations<T> = Partial<Record<Version, MigrationMap<any, T>>>;
```

Comment:
By far the biggest change. Before, migrations had to be:
- From the version of change to the most recent version (v1 -> v3). This means every time we added a new migration, we would have to go to those old versions and make sure they can properly be updated from say (v1 -> v4).  There was not concept of migrating one step at a time.
- The previous map had all migrations for all versions in one object. Was harder to understand (given we tend to think about migrations on a version 43 to version 44 kind of basis).

Reply:
resolving as outdated